new function: std.file.copyRecurse

aG0aep6G opened the pull request:
For when you need to copy a directory.

Possible issues:
* Timestamps and attributes are not copied for symlinks. See NOTEs in the code.
* Copying stops on the first error. No way to ignore errors and just copy as much as possible.
* Creation times are not preserved (Windows).
* Asserting that there are no symlinks on Windows, because `symlink` is Posix-only. But `attrIsSymlink` can apparently be true on Windows.

BBasile commented on std/file.d:
```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
+
+Throws: $(D FileException) on error.
+ */
+void copyRecurse(string from, string to,
```
I find the function usefull but I don't like the interface.

why not `copyDirectory(string from, string to, bool recursive = true, PreserveAttributes preserve = preserveAttributesDefault)` ?

jmdavis replied:
Why would recursive be optional? Just use `copy` if that's what you want. Copying a directory is recursive by its very nature; otherwise, you'd just be doing mkdir.

JakobOvrum replied:
The interface shouldn't require immutable characters in `from` and `to`

aG0aep6G replied:
> why not copyDirectory(string from, string to, bool recursive = true, PreserveAttributes preserve = preserveAttributesDefault) ?

I suppose with `recursive = false` this would copy files in the directory and create sub-directories, but it wouldn't step into the sub-directories and copy their contents, right?

Is there a need for that? As far as I'm aware, file managers don't supply that functionality. I don't see where one would need it.

jmdavis commented on std/file.d:
```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
+
+    immutable attrs = getLinkAttributes(from);
+
+    if (attrIsFile(attrs)) copy(from, to, preserve);
```
I'd very much like to see this work where `to` is a directory (and the documentation you wrote implies that it should). `copy` can't handle that. It insists that you list the file name at the end of the target rather than just a directory (which I find _very_ annoying). You can see a discussion on it in this bug report: https://issues.dlang.org/show_bug.cgi?id=3862. The conclusion of that was that we decided that it would probably be better to leave `copy` as-is and create a new function which copied similar to what you've done here (much as I think that `copy` is it stands is pretty useless). I still need to look over exactly how this implementation behaves, but what I'd like to see is a function that roughly does what `cp` does.

aG0aep6G replied:
You mean `copyRecurse("foo", "bar")` would create bar/foo when foo is a file and bar is a directory, right? I'm with Vladimir on this one. I don't think that's good default behaviour for a copying function (in a programming context). I think it would be surprising when after `copyRecurse("foo", "bar")` bar is not the same as foo.

jmdavis replied:
I truly don't understand that. Having to append the file name to the target is _very_ annoying. I never do it with `cp` unless I want a different file name, and from what I've seen, that's the case with most people. And if I'm programming instead of using the command line, I want the same behavior. Anything else is just extra work that makes my life harder. I screw up `std.file.copy` just about every time that I use it and get ticked off at it every time. Personally, what I want is pretty much literally `cp` without having to worry about the command line interpreting file names or having it give an error, because the file list is longer that it will handle. I was meaning to add a function like this that acts that way but hadn't gotten around to it yet. Anything else borders on useless IMHO. It's just too annoying to use.

CyberShadow replied:
> And if I'm programming instead of using the command line, I want the same behavior.

That we've been using a scripting language (shell/batch) that uses the same ideas both for direct user interaction and automation is a gigantic historical mistake - just look at all the unintended `rm -rf /` happening in scripts because of missed escaping or unset variables.

> Anything else is just extra work that makes my life harder.

Sometimes making life harder is good, because if you make things too easy, it can make life VERY DIFFICULT when things go wrong.

By the same logic you would remove the `r` and `f` switches to `rm`, because they are redundant after all, aren't they?

jmdavis replied:
Honestly, I wouldn't mind that, and that's what `std.file.rmdirRecurse` already does. `rm -rf /` is a problem, because it's stupidly legal even though there is no reason to do it ever. I rather like the Solaris folks' solution to that: they made it illegal (while sticking to the POSIX standard no less).

CyberShadow replied:
If you're OK with having different functions for `remove`, `rmdir` and `rmdirRecurse`, I don't understand why you're not OK with different functions for `copy` and `copyTo`.

jmdavis replied:
I'd be perfectly fine with a single function for removing, and I'm fine with multiple functions given that one of them does remove everything. Similarly, if we want to have multiple copy functions, just so long as one of them works like cp and will take a target directory, I'm fine with it, much as I think that the other is too annoying to be useful. However, I don't know why `copyRecurse` would require a file as the target rather than accepting a directory. But then again, I'd expect `copyRecurse` to function like `cp` does, and looking over the unit tests, it doesn't look like it does.

What I want is basically `cp` but without having to call to the shell and deal with escaping file names or any other quirks of the shell. If you guys want to have other functions that are much pickier, then fine, but personally, I have no interest in ever using them and think that they're a waste of space.

CyberShadow replied:
I thought (maybe I misremember) that we agreed that `copyTo` would *only* work if the target is a directory. As I understand, you want it to work both if the target doesn't exist or if it's a directory.

Such a function would be one which has a poorly defined contract, i.e. it's doing heuristics on your behalf. Code that relies on heuristics doesn't know what it's doing. I sure hope you don't plan to submit code that doesn't know what it's doing to Phobos or the D project in general.

jmdavis commented:
> Asserting that there are no symlinks on Windows, because symlink is Posix-only. But attrIsSymlink can apparently be true on Windows.

Windows does have symlinks now (starting with Vista, I believe - but we specifically support Windows 7 and newer). I'm not very familiar with how they work exactly, but a quick search shows a number of articles on them. `std.file.symlink` and `std.file.readLink` should probably be updated to support Windows (and that may be necessary for `copyRecurse` to work properly with symlinks on Windows), but in principle, `copyRecurse` should be able to work with symlinks on Windows.

jmdavis commented:
> Copying stops on the first error. No way to ignore errors and just copy as much as possible.

I'm not sure that it's critical for a first round of this (and we do a poor job with handling errors like this in general in std.file - e.g.  with `dirEntries`), but we really should have a way to handle errors with a function like `copyRecurse`, or it becomes useless outside of small cases. I don't know what the best way to deal with it would be, but the first thought that comes to mind would be a delegate which was called when an error occurred so that it could decide what to do. A default implementation which does not have any error-handling would still be desirable though and is effectively what you have now.

JakobOvrum commented on std/file.d:
```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
+    if (attrIsFile(attrs)) copy(from, to, preserve);
+    else if (attrIsDir(attrs))
+    {
+        mkdirRecurse(std.conv.to!string(to));
```
`to` here looks like a noop

JakobOvrum replied:
Is there any reason this uses `mkdirRecurse` instead of `mkdir` every time? Shouldn't it do the extra logic just for the root target directory?

aG0aep6G replied:
> to here looks like a noop

Oh yeah, that's a left-over from when I tried to support ranges. Fixed.

> Is there any reason this uses mkdirRecurse instead of mkdir every time? Shouldn't it do the extra logic just for the root target directory?

mkdir throws when the directory exists; mkdirRecurse doesn't. Could replace with `if (!exists(to)) mkdir(to);`, I guess.

JakobOvrum commented on std/file.d:
```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
+        
+        SysTime accessTime, modificationTime;
+        getTimes(from, accessTime, modificationTime);
+        setTimes(to, accessTime, modificationTime);
```
So when copying fails, the associated time information is different from a successful copy? Could this be avoided by rearranging this?

aG0aep6G replied:
Copying the contents alters the directory's times. So updating the directory times has to happen after copying the contents. Could `try {copy contents} catch(FileException e) finally {update times; throw e;}`, I guess. But is there a point in going out of our way to ensure preservation of times when copying itself has failed?

JakobOvrum commented on std/file.d:
```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
+                ones, for now. */
+                if (getLinkAttributes(to) != attrs)
+                {
+                    throw new FileException(from, "Cannot preserve " ~
```
Use `enforce`

aG0aep6G replied:
done

JakobOvrum commented:
This function allocates *a lot* of short-lived GC memory, one chunk for each unique path in the tree. This can and should be avoided.

A related issue is that it should ideally work with all range paths, not just array paths. This enhancement has been done for `copy` and much of Phobos. This helps with pushing allocation decisions downward.

I suspect this approach gives suboptimal performance, especially when `from` and `to` are on different physical devices, but this could be measured. Instead of trying to reimplement `rsync` in the standard library, we could simply warn about this in the documentation and suggest that users invoke `rsync` (or whatever) instead.

CyberShadow commented on std/file.d:
```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
+{
+    // Remove any trailing directory separators.
+    from = buildNormalizedPath(from);
+    to = buildNormalizedPath(to);
```
std.path needs Delphi's `ExcludeTrailingPathDelimiter`.

aG0aep6G replied:
Agreed. In the meantime, I should probably replace buildNormalizedPath with something that only checks for the trailing separator.

CyberShadow commented on std/file.d:
```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
+                /*NOTE: Would copy attributes of the symlink here, but there
+                is no setLinkAttributes here, and no fchmodat in the C headers.
+                Just enforcing that the new attributes are the same as the old
+                ones, for now. */
```
I'm not sure we should settle on any user-visible behavior "for now", because then fixing it becomes a breaking change.

aG0aep6G replied:
Fixing it would mean that a call to copyRecurse then properly copies everything instead of throwing an exception. Depending on that particular exception would seem silly to me. But, yeah, I guess it's technically a breaking change.

We could settle with the throwing behaviour for good, of course. Or maybe it's possible to implement copying of symlink attributes. I'm not so well informed on Linux/Posix/Unix stuff, but as far as I could find out, fchmodat is needed to do it. `man 2 fchmodat` says that it "was added to Linux in kernel 2.6.16; library support was added to glibc in version 2.4.", and that it's in POSIX.1-2008. That seems rather new. I don't know if it's too new for us to use.

Also note that most systems apparently don't even have a way to change the permissions of symlinks. OS X seems to be the outlier in that it does not have fixed symlink permissions.

aG0aep6G commented:
> copyRecurse should be able to work with symlinks on Windows.

Agreed.

> we really should have a way to handle errors with a function like copyRecurse, or it becomes useless outside of small cases. I don't know what the best way to deal with it would be, but the first thought that comes to mind would be a delegate which was called when an error occurred so that it could decide what to do.

I'd been playing around with a version that collects all thrown exceptions and throws a single one at the end of the operation. I.e., it would copy as much as possible and then inform the caller via exception about what failed. The delegate thing would be more flexible, so that sounds good to me.

> This function allocates a lot of short-lived GC memory, one chunk for each unique path in the tree. This can and should be avoided.
>
> A related issue is that it should ideally work with all range paths, not just array paths. This enhancement has been done for copy and much of Phobos. This helps with pushing allocation decisions downward.

I had tried that, but got frustrated when some of std.file's functions would take ranges but others only const arrays (mkdir vs mkdirRecurse). dirEntries even demands a string. I'll try and reduce the allocations. Any non-obvious pointers?

> I suspect this approach gives suboptimal performance, especially when from and to are on different physical devices, but this could be measured. Instead of trying to reimplement rsync in the standard library, we could simply warn about this in the documentation and suggest that users invoke rsync (or whatever) instead when performance is desired.

You mean we could try and skip files that haven't changed? I think we shouldn't try to be that clever here. Keep it simple, just do what the name says: copy everything.

aG0aep6G commented:
I'm a little confused about the changelog. I had to rebase and new changelog entries showed up. They seem to be the ones at <http://dlang.org/changelog/2.070.0.html>. Shouldn't the changelog for 2.070 be in the stable branch now?

JakobOvrum commented:
> I had tried that, but got frustrated when some of std.file's functions would take ranges but others only const arrays (mkdir vs mkdirRecurse). dirEntries even demands a string. I'll try and reduce the allocations. Any non-obvious pointers?

Those functions have yet to be improved. `dirEntries` is the only tough one, but one we have to solve to get `@nogc` going.

> You mean we could try and skip files that haven't changed?

No, a good tool reads and writes asynchronously.

JakobOvrum commented on std/file.d:
```diff
@@ -3200,6 +3200,217 @@ version(Posix) unittest //issue 11434
+        {
+            if (exists(to)) remove(to);
+            symlink(readLink(from), to);
+            if (preserve == PreserveAttributes.yes)
```
Just collapse this if-statement into the enforce condition

aG0aep6G commented:
Somewhat larger update. New commits start at c178944 ("remove @safe from template setTimes").

Summary:
* Rangified `readLink`, `symlink`, and (more or less) `copyRecurse`.
* Removed `@safe` from `setTimes`, `readLink`, and `symlink`. Attribute reference can take care of that.
* Fixed wrong documentation of `symlink`.

There are still two instances of stringification in `copyRecurse` (aside from when throwing exceptions): `mkdirRecurse` and `dirEntries`. I'm afraid I might not have the will power to rangify those.

Overall, I'm not very confident in my rangification of `copyRecurse`. If it's rubbish, please advise.

Other open issues:

* Symlinks on Windows - Since we don't support anything older than Windows 7, and Windows 7 supports symlinks, we should support symlinks on Windows. I couldn't find any declaration of [`CreateSymbolicLink`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363866%28v=vs.85%29.aspx), though. What do I do here?

* Attributes and timestamps of symlinks on Posix. Similar situation to the Windows thing, but it's less clear if we can support it or if the needed Posix functions are too new. Those would be `fchmodat` and `lutimes`, as far as I could google.

By the way, I've noticed that `std.internal.cstring.tempCString` is unsafe and @trusted. Not cool.

ZombineDev commented:
@aG0aep6G `CreateSymbolicLink` should be included in [winbase.d](https://github.com/D-Programming-Language/druntime/blob/master/src/core/sys/windows/winbase.d), but it looks like druntime has an old version of the Windows SDK headers and we will need to update it to include APIs supported by newer version of Windows.

CyberShadow commented:
On Windows, we should create junctions (not symlinks) for directories, as that doesn't require elevated privileges by default.

Here's my code to create junctions and symlinks, feel free to submit to std.file:

https://github.com/CyberShadow/ae/blob/b8fc2b12a99f4554b8822938d8c548abfff33f5e/sys/file.d#L556-L658

aG0aep6G commented:
I've split off the prerequisite parts of, and made individual pull requests:

*  Cleanup changelog #3964
*  remove @safe from template setTimes #3965
*  fix documentation for symlink #3966
*  rangify readLink and symlink #3967

DmitryOlshansky commented:
@aG0aep6G needs rebase

aG0aep6G commented:
> @aG0aep6G needs rebase

This needs more than a rebase. For a start, it needs #3967.

wilzbach commented:
> This needs more than a rebase. For a start, it needs #3967.

aG0aep6G Your wish has been heard - wanna rebase?  ;-)

aG0aep6G commented:
> aG0aep6G Your wish has been heard - wanna rebase? ;-)

Rebased, but I don't think this is ready to be merged.

DmitryOlshansky commented:
> Rebased, but I don't think this is ready to be merged.

Needs a new rebase, what else do you have on your mind?

aG0aep6G commented:
> what else do you have on your mind?

1. Symlinks on Windows

    I just tried to add `CreateSymbolicLink` to `winbase.d` [here](https://github.com/aG0aep6G/druntime/commit/13ae8ca097f9be55102635982b89e061ee511bad), but either I messed up the declaration, or it's not in dmd's kernel32.lib. Giving up for today. Help would be appreciated.

    CyberShadow has provided [some code](https://github.com/CyberShadow/ae/blob/b8fc2b12a99f4554b8822938d8c548abfff33f5e/sys/file.d#L556-L658), but it looks rather daunting. Symlinks, junction points, reparse points ... I'd need time and motivation to educate myself on this Windows stuff.

    Of course we could also decide that we don't need to support symlinks on Windows for now.

2. Attributes of symlinks on Posix

    I don't know if we can rely on the needed functions `fchmodat` and `lutimes`. Again, not supporting this is an option as well.

3. Maybe try and cut down more on allocations.

    `text` is called in the calls to `mkdirRecurse` (only root directory) and `dirEntries` (every directory). Could maybe rangify those.

    Multiple functions are called on `src.save`/`dst.save` of which each calls `tempCString`. Could maybe call `tempCString` once and pass the result.

andralex commented:
I'm okay with the addition assuming some non-tenuous semantics can be defined. There are some implementation issues that I'd like to see improved.

wilzbach commented:
> Rebased, but I don't think this is ready to be merged.

@aG0aep6G so I think this PR would be a lot more important than the `CHEATSHEET` one.
Do you have some time to finish it up?

aG0aep6G commented:
> @aG0aep6G so I think this PR would be a lot more important than the CHEATSHEET one.
> Do you have some time to finish it up?

The issues above still stand. If someone can give me directions on those, sure, I can work on this. Otherwise I'm willing to admit defeat here.

MetaLang commented:
@aG0aep6G are you still working on this or can I close it?
